### PR TITLE
[Fix] LmdbBackend can't pickle env obj

### PR DESCRIPTION
**Pull request by cuhk-hbsun**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fix bug: can't pickle environment objects (when multiprocess start method is spawn)
see [link](https://github.com/pytorch/examples/issues/526)

## Modification

- Update mmcv/fileio/file_client.py

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/fileio/file_client.py [hunk 1]
```diff
@@ -473,33 +473,47 @@ def __init__(self,
...
 
     def get_text(self, filepath, encoding=None):
         raise NotImplementedError
 
+    def __getitem__(self, index):
+        if not hasattr(self, 'env'):
+            self.env = self._get_env()
+
+        with self.env.begin(write=False) as txn:
+            value_buf = txn.get(str(index).encode('utf-8'))
```
[the remark this answers is not in the document]
Comment (cuhk-hbsun) [reply mmcv/fileio/file_client.py:498]: 1. For the index which is a number (e.g. 6) or English string (e.g. sample.jpg), `str(index).encode('ascii') == str(index).encode('utf-8')`, so no BC-breaking.
2. For the index which contains Chinese character (e.g. 样本.jpg), `str(index).encode('ascii')` will give error 
"UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-1: ordinal not in range(128)"

**Review by AllentDan — APPROVED** [review]
LGTM

**Comment by HAOCHENYE** [comment]
LGTM.

**Review by HAOCHENYE — APPROVED** [review]

**Comment by ZwwWayne** [comment]
Hi @cuhk-hbsun ,
Could you resolve the lint issue so that we can merge this PR?

File: mmcv/fileio/file_client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZwwWayne) [reply mmcv/fileio/file_client.py:498]: Great! Then I suggest we leave comments here and refer to this issue as a reminder for future developers.

File: mmcv/fileio/file_client.py [hunk 2]
```diff
@@ -473,33 +473,47 @@ def __init__(self,
...
+
+        return lmdb.open(
+            self.db_path,
+            readonly=self.readonly,
+            lock=self.lock,
+            readahead=self.readahead,
+            **self.kwargs)
+
+    def close(self):
+        self.env.close()
```
Comment (zhouzaida) [inline mmcv/fileio/file_client.py:515]: Should we close env in `__del__`  rather than `close` because other backends do have not this `close` method?

Comment (cuhk-hbsun) [reply mmcv/fileio/file_client.py:515]: Let me have a try. `env.close()` is only useful for pytest on Windows and can be delete here.

**Review by gaotongxiao — APPROVED** [review]

File: mmcv/fileio/file_client.py [hunk 3]
```diff
@@ -473,33 +473,47 @@ def __init__(self,
...
+
+        with self.env.begin(write=False) as txn:
+            value_buf = txn.get(str(index).encode('utf-8'))
+        return value_buf
+
+    def _get_env(self):
+        try:
+            import lmdb
+        except ImportError:
+            raise ImportError('Please install lmdb to enable LmdbBackend.')
```
Comment (zhouzaida) [inline mmcv/fileio/file_client.py:505]: Can we move this `try import` to `__init__` and just `import lmdb` in `_get_env`?

**Review by zhouzaida — APPROVED** [review]

File: mmcv/fileio/file_client.py [hunk 4]
```diff
@@ -474,32 +474,45 @@ def __init__(self,
...
-        self._client = lmdb.open(
-            self.db_path,
-            readonly=readonly,
-            lock=lock,
-            readahead=readahead,
-            **kwargs)
+        self.readonly = readonly
+        self.lock = lock
+        self.readahead = readahead
+        self.kwargs = kwargs
```
Comment (zhouzaida) [inline mmcv/fileio/file_client.py:485]:
```suggestion
        self.kwargs = kwargs
        self._client = None
```

File: mmcv/fileio/file_client.py [hunk 5]
```diff
@@ -474,32 +474,45 @@ def __init__(self,
...
         """Get values according to the filepath.
 
         Args:
             filepath (str | obj:`Path`): Here, filepath is the lmdb key.
         """
-        filepath = str(filepath)
-        with self._client.begin(write=False) as txn:
-            value_buf = txn.get(filepath.encode('ascii'))
+        if not hasattr(self, 'env'):
+            self.env = self._get_env()
```
Comment (zhouzaida) [inline mmcv/fileio/file_client.py:494]:
```suggestion
        if self._client is None:
            self._client = self._get_client()
```

File: mmcv/fileio/file_client.py [hunk 6]
```diff
@@ -474,32 +474,45 @@ def __init__(self,
...
+
+        with self.env.begin(write=False) as txn:
+            value_buf = txn.get(str(filepath).encode('utf-8'))
         return value_buf
 
     def get_text(self, filepath, encoding=None):
         raise NotImplementedError
 
+    def _get_env(self):
```
Comment (zhouzaida) [inline mmcv/fileio/file_client.py:503]:
```suggestion
    def _get_client(self):
```

File: mmcv/fileio/file_client.py [hunk 7]
```diff
@@ -474,32 +474,45 @@ def __init__(self,
...
+
+        return lmdb.open(
+            self.db_path,
+            readonly=self.readonly,
+            lock=self.lock,
+            readahead=self.readahead,
+            **self.kwargs)
+
+    def __del__(self):
+        self.env.close()
```
Comment (zhouzaida) [inline mmcv/fileio/file_client.py:514]:
```suggestion
        self._client.close()
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**